File: server/server_advanced/student_context.py
import time


def student_context(state, now_ms=None):
    user = state.get('user') or {}
    position = user.get('currentPosition') or {}
    if position.get('posTypeCode') != 'P':
        return {'prsId': user.get('prsId'), 'userId': state.get('userId') or user.get('userId')}

    children = [child for child in position.get('myChildren', [])
                if isinstance(child, dict) and type(child.get('prsId')) is int and child['prsId'] > 0]
    if not children:
        raise ValueError('В родительском аккаунте нет доступного ребёнка')
    child = next((child for child in children if child.get('isDefaultChild')), children[0])
    now_ms = time.time() * 1000 if now_ms is None else now_ms
    enrollments = [item for item in child.get('userData', [])
                   if isinstance(item, dict) and item.get('orgIsReady') and item.get('userId')]
    current = [item for item in enrollments
               if isinstance(item.get('fullStartDate'), (int, float))
               and isinstance(item.get('fullEndDate'), (int, float))
               and item['fullStartDate'] <= now_ms <= item['fullEndDate']]
    enrollment = next(iter(current), None) or next(
        (item for item in enrollments if item.get('yearState') == 'CURR'), None)
    if enrollment is None and enrollments:
        enrollment = max(enrollments, key=lambda item: item.get('fullStartDate') or 0)
    return {'prsId': child['prsId'], 'userId': (enrollment or {}).get('userId')}

File: lib/models/student_context.dart
/// владелец родительской сессии и ученик имеют разные идентификаторы
class StudentContext {
  final int? prsId;
  final int? userId;

  const StudentContext({this.prsId, this.userId});

  factory StudentContext.fromState(
    Map<String, dynamic> state, {
    DateTime? now,
  }) {
    final user = state['user'] as Map? ?? const {};
    final position = user['currentPosition'] as Map? ?? const {};
    if (position['posTypeCode'] != 'P') {
      return StudentContext(
        prsId: user['prsId'] as int?,
        userId: (state['userId'] ?? user['userId']) as int?,
      );
    }
    final children = (position['myChildren'] as List? ?? const [])
        .whereType<Map>()
        .where((child) => child['prsId'] is int && child['prsId'] > 0)
        .toList();
    if (children.isEmpty) return const StudentContext();
    final child = children.firstWhere(
      (child) =>
          child['isDefaultChild'] == true || child['isDefaultChild'] == 1,
      orElse: () => children.first,
    );
    final timestamp = (now ?? DateTime.now()).millisecondsSinceEpoch;
    final enrollments = (child['userData'] as List? ?? const [])
        .whereType<Map>()
        .where(
          (item) =>
              (item['orgIsReady'] == true || item['orgIsReady'] == 1) &&
              item['userId'] is int &&
              item['userId'] > 0,
        )
        .toList();
    final current = enrollments
        .where(
          (item) =>
              item['fullStartDate'] is num &&
              item['fullEndDate'] is num &&
              item['fullStartDate'] <= timestamp &&
              timestamp <= item['fullEndDate'],
        )
        .toList();
    final active = enrollments
        .where((item) => item['yearState'] == 'CURR')
        .toList();
    enrollments.sort(
      (a, b) => ((b['fullStartDate'] as num?) ?? 0).compareTo(
        (a['fullStartDate'] as num?) ?? 0,
      ),
    );
    final candidates = current.isNotEmpty
        ? current
        : active.isNotEmpty
        ? active
        : enrollments;
    return StudentContext(
      prsId: child['prsId'] as int,
      userId: candidates.isEmpty ? null : candidates.first['userId'] as int,
    );
  }
}
